src/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.HandleObject.cs

```diff
@@ -47,6 +71,23 @@ private static bool HandleEndObject(JsonSerializerOptions options, ref ReadStack
 
             state.Current.JsonClassInfo.UpdateSortedPropertyCache(ref state.Current);
 
+            if (state.Current.IsDictionary())
+            {
+                ReadStackFrame frame = state.Current;
+
+                int keyCount = ((IList)frame.TempDictKeys).Count;
+
+                Debug.Assert(keyCount == (frame.TempDictValues).Count);
+
+                IList keys = frame.TempDictKeys;
+                IList values = frame.TempDictValues;
+
+                for (int i = 0; i < keyCount; i++)
+                {
+                    ((IDictionary)state.Current.ReturnValue).Add(keys[i], values[i]);
```

Comment:
This copying from list to dictionary is surprising. @steveharter, is this expected?